feat: llm-gateway: anthropic models for fallbacks [addresses ENT-1391] (#5196)
langchain-ai/docs: src/langsmith/llm-gateway-fallbacks.mdx

For each request, the gateway:

1. Determines the wire format from the request path, and considers only the chains in that format (see [Wire formats](#wire-formats)).
1. Picks one of those chains (see [Selecting a chain](#selecting-a-chain)).
1. Calls the chain's first model configuration.
1. If the response status matches a configured trigger, discards that response and calls the next model configuration in the chain.
1. Repeats until a candidate returns a non-trigger response, or the chain is exhausted—in which case the last candidate's response is returned to the caller.

> Each model configuration in a chain can point to a different provider and model identifier, and the gateway calls every candidate with that configuration's own model name in place of the one the client sent—on the first attempt as well as on any fallback. The request's `model` serves only to pick a chain among those matching the path's [wire format](#wire-formats); if it matches none of them, the gateway uses that format's first (default) chain.

Reviewer: ```suggestion
Each configuration in a chain can point to a different provider and model. On both the first attempt and any fallback, the gateway calls the candidate with its configured model name, replacing the value the client sent. The client's model field only selects which chain to use from those matching the path's [wire format](#wire-formats). If no chain matches, the gateway uses the first chain defined for that wire format as the default.
```

## Wire formats

Each chain is either OpenAI-compatible or Anthropic-compatible, since the gateway forwards the same request body to every candidate in it:

| Chain format | Model configuration provider | Path on `/routes/{name}` |
| --- | --- | --- |
| OpenAI-compatible | **OpenAI Compatible Endpoint** | `POST /v1/chat/completions`, `POST /v1/responses` |
| Anthropic-compatible | **Anthropic** | `POST /v1/messages` |

> Only these two provider types can go in a chain. A [model configuration](/langsmith/model-configurations) saved for another provider, such as Azure OpenAI or Bedrock, isn't eligible—to reach a host that speaks the OpenAI API, save it as an **OpenAI Compatible Endpoint** with its base URL.

Reviewer: ```suggestion
Only these two provider types can go in a chain. A [model configuration](/langsmith/model-configurations) saved for another provider, such as Azure OpenAI or Bedrock, isn't eligible. To reach a host that speaks the OpenAI API, save it as an **OpenAI Compatible Endpoint** with its base URL.
```

The path you call selects the format, and the gateway only considers chains in that format. Any other path returns `501 Not Implemented`.

> A single fallback configuration can hold chains of both formats, so one route can serve both OpenAI-compatible and Anthropic-compatible clients. It doesn't have to: a configuration with only OpenAI-compatible chains returns `502` for a `/v1/messages` call, and one with only Anthropic-compatible chains returns `502` for a `/v1/chat/completions` call.

Reviewer: ```suggestion
A single fallback configuration can hold chains of both wire formats, so one route can serve both OpenAI-compatible and Anthropic-compatible clients. Mixing formats is optional: a configuration with only OpenAI-compatible chains returns `502` for `/v1/messages` calls, and one with only Anthropic-compatible chains returns `502` for `/v1/chat/completions` calls.
```

1. Click **Create configuration**.
1. Enter a **Configuration name**. This becomes `{name}` in the gateway URL `https://gateway.smith.langchain.com/routes/{name}`.
1. Select the **Workspace** the configuration belongs to. [Model configurations](/langsmith/model-configurations) are workspace-scoped, so only that workspace's configurations are available to add to a chain. The workspace can't be changed later—delete and recreate the configuration to move it.
1. Under **Fallback triggers**, review the HTTP status codes that should trigger a fallback. The list comes prepopulated with the transient codes another provider has a chance of serving (such as `429`, `500`, and `503`); add or remove codes as needed.

> 1. Under **Model fallback chains**, click **Add chain**, then add two to five model configurations to it in the order they should be tried. A chain's first model fixes its [wire format](#wire-formats), and only configurations of that format can be added after it. Chains are grouped by format, and the first chain in each group is that format's **default** chain, used whenever a request's model doesn't select a different one.

Reviewer: ```suggestion
1. Under **Model fallback chains**, click **Add chain**, then add two to five model configurations in the order the gateway should try them. A chain's first model fixes its [wire format](#wire-formats); only configurations of that format can follow. The gateway groups chains by format, and the first chain in each group is that format's **default**, which it uses when a request's model does not select another chain.
```

```

```bash Anthropic-compatible
curl https://gateway.smith.langchain.com/routes/my-route/v1/messages \
    -H "Authorization: Bearer $LANGSMITH_API_KEY" \
    -H "Content-Type: application/json" \
    -d '{"max_tokens":1024,"messages":[{"role":"user","content":"ping"}]}'
```

</CodeGroup>

The gateway tries each model configuration in the selected chain, in order, until one responds without a trigger status. Each attempt is traced and counted against [spend policies](/langsmith/llm-gateway-spend-policies) on its own, so a request that falls back records one call per candidate tried.

## Selecting a chain

A configuration can hold more than one fallback chain, which is useful when different model families need different fallback behavior. Among the chains matching the request's [wire format](#wire-formats), the gateway selects one by the request body's `model` field, in this order:

1. If `model` matches a chain's **alias**, that chain is used.
1. Otherwise, if `model` matches a chain's **primary** (first) model configuration's underlying model name, that chain is used.
1. If `model` is omitted, or matches neither, the first (default) chain of that format is used.

> An alias is optional and set per chain when you create the configuration. It's a caller-facing name, so a client can ask for `"model": "heavy"` without knowing which model backs it.

Reviewer: ```suggestion
An alias is optional and set per chain when you create the configuration. It is a caller-facing name, so a client can ask for `"model": "heavy"` without knowing which model backs it.
```

> Each model configuration in a chain can point to a different host. So a chain can fail over across separate deployments of the same model (for example, a primary endpoint to a backup endpoint) without a single host being a point of failure.

Reviewer: ```suggestion
Each model configuration in a chain can point to a different host, so a chain can fail over across separate deployments of the same model (for example, from a primary endpoint to a backup) with no single host as a point of failure.
```

For example, a configuration with three chains:

| Chain | Format | Models (in priority order) |
| --- | --- | --- |
| 1 (default OpenAI-compatible) | OpenAI-compatible | `gpt-5.5` on OpenAI → `gpt-5.5` on Azure OpenAI → `llama-3.3-70b` on a self-hosted endpoint |
| 2 | OpenAI-compatible | `gpt-4o-mini` on OpenAI → `kimi-k2` on Fireworks |
| 3 (default Anthropic-compatible) | Anthropic-compatible | `claude-sonnet-4-6` on Anthropic → `claude-haiku-4-5` on Anthropic |

- A `/v1/chat/completions` request with `"model": "gpt-5.5"` uses chain 1, which fails over from OpenAI to Azure OpenAI to the self-hosted endpoint.
- A `/v1/chat/completions` request with `"model": "gpt-4o-mini"` uses chain 2, which fails over from OpenAI to Fireworks.
- A `/v1/chat/completions` request with an unrecognized or omitted model uses chain 1, the default for that format.

> - A `/v1/messages` request uses chain 3 whatever its `model` is, since it's the only Anthropic-compatible chain—chains 1 and 2 are never eligible on that path.

Reviewer: ```suggestion
- A `/v1/messages` request uses chain 3 whatever its `model` is, because it's the only Anthropic-compatible chain. Chains 1 and 2 are never eligible on that path.
```
